### DISCUSSION TITLE
Possibility to specify Switch state ON/OFF when Wifi Signal is lost

### DISCUSSION THREAD

**Discussion by berndh1979** [arendst/Tasmota#20225] [Ideas]
I have searched if this feature is already included, but can't find it. 

I have a tasmota switch (ShellyPlug S flashed). It work nicely, however when the Wifi signal gets lost I can't control the device anymore from HA obviously. However the state of the swtich is important. It should switch ON to make sure that the device connected to this switch keeps running. So I was wondering if there is an option available to specify the state of the switch in case the wifi signal is lost. 

Thanks

**Comment by barbudor** [comment]
In Tasmota, a "Switch" is a input signal tight to the state to the external signal it is attached to. You can't change it
Alternatively, a Relay is an output that you can control through the `Power` command

So you could define what we call a 'virtual relay' connected to a free gpio and use rules triggered by 'on wifi#connected' and 'on wifi#disconnected' to change the power state. That change will then be reported

**Reply by barbudor** [reply]
Stupid me, if the wifi is disconnected, how would the state be reported to HA?

**Reply by barbudor** [reply]
My bad, it's not about reporting is exactly about enforcing power on when wifi gets disconnected?

So you have the answer above, no need for a virtual switch, just use 'on wifi#disconnected do power on endon'

**Comment by berndh1979** [comment]
Thanks, indeed I'm not interested in the reporting to HA if the wifi get's disconnected... ;)

But I did not know I can make rules in the tasmota device it self. I will look into that. Thanks

**Comment by berndh1979** [comment]
A quick look into the syntax:

Rule1 ON Wifi#Disconnected DO Power On ENDON <enter>
Rule1 1 <enter>

The first line is the Rule1 syntax to do exactly what I would like and line number 2 actually enables Rule1.

Feedback that I get in the terminal:
stat/tasmota_EA9783/RESULT = {"Rule1":{"State":"ON","Once":"OFF","StopOnError":"OFF","Length":38,"Free":473,"Rules":"ON Wifi#Disconnected DO Power On ENDON"}}
13:19:52.717 RUL: WIFI#DISCONNECTED performs "Power On"
13:19:52.724 MQT: stat/tasmota_EA9783/RESULT = {"POWER":"ON"}
13:19:52.729 MQT: stat/tasmota_EA9783/POWER = ON

Seems now that the device is actually switched on, while it is actively connected to Wifi. 

So I'm lost here. Why does the Rule1 enables the switch, while it is connected to WIfi?

**Reply by berndh1979** [reply]
Hmmmm, when I do the following:

Rule1 0
Rule1 1 

It enables the rule again, but this time it did not switch the power switch on... 

Aparently in the initial case it actually disconnected the Wifi and confirms this with the statement:

13:19:52.717 RUL: WIFI#DISCONNECTED performs "Power On"

and hence it switched Power ON. It did not do that in the second enablement attempt....

Was I just very unlucky that the two (enable rule AND Wifi#Disconnected) actually happend at the exact same time... or is there an explanation for this behavior.

**Comment by berndh1979** [comment]
So, the Wifi#Disconnected now seems to work nicely. Jeehaa, thanks for putting me into the right direction. 

It would also be nice when the wifi is restored that it actually takes over the state that was last reported prior to losing the Wifi signal. 

The actual state I would like to have it, is stored in MQTT as HA writes the state into MQTT. So when the signal get's restored it should read the last MQTT state message. 

Any suggestions there?

**Reply by barbudor** [reply]
So you need to store the previous state either into a VAR and be sure to have this VAR set properly at boot and in case of change:
```
on Power1#boot do var1 %value% endon
on Power1#state do var1 %value% endon
``` 
Then on wifi disconnection, copy the current value to a 2nd var before changing power state:
```
on wifi#disconnected do backlog var2 %var1%; power on endon
```
And finally restore the state:
```
on wifi#connected do power %var2% endon
```

All individual rules above can be stored inside `Rule1`, just copy them one after the other with a space in between

**Reply by berndh1979** [reply]
Will give it a try. Again thanks for putting me into the right direction.

**Reply by berndh1979** [reply]
This seems to work nicely for my application. I guess after the wifi#connected it will read the mqtt queue again and switch the state based upon the mqtt messages pending in the queue if there are any, right?

**Reply by sfromis** [reply]
Generally, there is no "queue" of MQTT messages, as there is no concept of sending to some specific device. Instead messages are made available based on active subscriptions. However there can be retained messages getting delivered after each subscription to matching topics, typically after connection.
